census/core.py

```diff
@@ -187,10 +230,31 @@ def query(self, fields, geo, year=None, **kwargs):
 
             headers = data.pop(0)
             types = [self._field_type(header, year) for header in headers]
-            results = [{header : (cast(item) if item is not None else None)
-                        for header, cast, item
-                        in zip(headers, types, d)}
-                       for d in data]
+            results = []
+            error = False
+            for d in data:
+                row = []
+                for header, cast, item in zip(headers, types, d):
+                    if item is not None:
+                        try:
+                            result = {header: cast(item, cast_nulls)}
+                        except NullValueException:
+                            # This value needs to raise an error, but we need the
+                            # rest of the row values for context, so flag the
+                            # error and continue the iteration
+                            error = True
+                            result = {header: item}
```

Comment:
Some additional context for why we need to finish this inner loop before raising the error is that the elements generated by `for d in data` look something like this:

```python
{'var_name': -666666666, 'state': 42, 'county': 11}
```

The inner loop `for header, cast, item in zip(headers, types, d)` processes each of these elements in turn, returning a formatted "row" for the final results. The problem with raising an error in the inner loop is that we'll only know one of these `header: item` pairs (like `var_name: -666666666` or `'state': 42`) but unless you know all of the elements it's impossible to determine which API call raised the error.